pkg/validation-webhook/snapshot.go

```diff
@@ -77,15 +94,59 @@ func admitSnapshot(ar v1.AdmissionReview) *v1.AdmissionResponse {
 			klog.Error(err)
 			return toV1AdmissionResponse(err)
 		}
-		return decideSnapshotContent(snapcontent, oldSnapcontent, isUpdate)
+		return decideSnapshotContentV1beta1(snapcontent, oldSnapcontent, isUpdate)
+	case SnapshotContentV1GVR:
+		snapcontent := &volumesnapshotv1.VolumeSnapshotContent{}
+		if _, _, err := deserializer.Decode(raw, nil, snapcontent); err != nil {
+			klog.Error(err)
+			return toV1AdmissionResponse(err)
+		}
+		oldSnapcontent := &volumesnapshotv1.VolumeSnapshotContent{}
+		if _, _, err := deserializer.Decode(oldRaw, nil, oldSnapcontent); err != nil {
+			klog.Error(err)
+			return toV1AdmissionResponse(err)
+		}
+		return decideSnapshotContentV1(snapcontent, oldSnapcontent, isUpdate)
 	default:
 		err := fmt.Errorf("expect resource to be %s or %s", SnapshotV1Beta1GVR, SnapshotContentV1Beta1GVR)
 		klog.Error(err)
 		return toV1AdmissionResponse(err)
 	}
 }
 
-func decideSnapshot(snapshot, oldSnapshot *volumesnapshotv1beta1.VolumeSnapshot, isUpdate bool) *v1.AdmissionResponse {
+func decideSnapshotV1beta1(snapshot, oldSnapshot *volumesnapshotv1beta1.VolumeSnapshot, isUpdate bool) *v1.AdmissionResponse {
+	reviewResponse := &v1.AdmissionResponse{
+		Allowed: true,
+		Result:  &metav1.Status{},
+	}
+
+	if isUpdate {
+		// if it is an UPDATE and oldSnapshot is not valid, then don't enforce strict validation
+		// This allows no-op updates to occur on snapshot resources which fail strict validation
+		// Which allows the remover of finalizers and therefore deletion of this object
+		// Don't rely on the pointers to be nil, because the deserialization method will convert it to
+		// The empty struct value. Instead check the operation type.
+		if err := utils.ValidateV1Beta1Snapshot(oldSnapshot); err != nil {
+			return reviewResponse
+		}
+
+		// if it is an UPDATE and oldSnapshot is valid, check immutable fields
+		if err := checkSnapshotImmutableFieldsV1beta1(snapshot, oldSnapshot); err != nil {
+			reviewResponse.Allowed = false
+			reviewResponse.Result.Message = err.Error()
+			return reviewResponse
+		}
+	}
+	// Enforce strict validation for CREATE requests. Immutable checks don't apply for CREATE requests.
+	// Enforce strict validation for UPDATE requests where old is valid and passes immutability check.
+	if err := utils.ValidateV1Beta1Snapshot(snapshot); err != nil {
+		reviewResponse.Allowed = false
+		reviewResponse.Result.Message = err.Error()
+	}
+	return reviewResponse
+}
+
+func decideSnapshotV1(snapshot, oldSnapshot *volumesnapshotv1.VolumeSnapshot, isUpdate bool) *v1.AdmissionResponse {
```

Comment:
for v1 we don't need ratcheting validation check right? we should always check for immutability

Reply:
For v1, the webhook does not check if there is only one source as that is validated by the schema already.
In utils.ValidateSnapshot, we do have a check to make sure the VolumeSnapshotClassName is not an empty string.

Reply:
Yes, I meant, can we remove the block L155 below that checks if the original object is validate on an update?

Reply:
Done

Reply:
Thanks! will let @yuxiangqian provide final review